File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def check_melody(play_time, melody, m):
    m_length = len(m)
    result = False
    for i in range(play_time - m_length + 1):
        for j in range(i, i + m_length):
            if m[j - i] != melody[j % len(melody)]:
                break
        else:
            result = True
            break
    return result


def solution(m, musicinfos):
    answer = ["(None)", 0]
    # 음#을 한 문자로 대체
    replace_dict = {"C#" : "c", "D#" : "d", "F#" : "f", "G#" : "g", "A#" : "a"}
    for k, v in replace_dict.items():
        m = m.replace(k, v)
    
    for musicinfo in musicinfos:
        start_time, end_time, title, melody = musicinfo.split(",")
        # 음#을 한 문자로 대체
        for k, v in replace_dict.items():
            melody = melody.replace(k, v)
        # 플레이시간 계산
        play_time = (int(end_time[0:2]) - int(start_time[0:2])) * 60 + (int(end_time[3:5]) - int(start_time[3:5]))
        # 동일 멜로디가 있고, 플레이 시간이 더 길면 answer 갱신
        if check_melody(play_time, melody, m) == True and answer[1] < play_time:
            answer = [title, play_time]
    
    return answer[0]

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
class Solution {
    
    public String flatReplace(String melody) {
        return melody.replace("C#", "c")
            .replace("D#", "d")
            .replace("F#", "f")
            .replace("G#", "g")
            .replace("B#", "b")
            .replace("A#", "a");
    }
    
    public String solution(String m, String[] musicinfos) {
        String answer = "(None)";
        int maxT = 0;
        m = flatReplace(m);
        
        for (String musicinfo : musicinfos) {
            String[] info = musicinfo.split(",");
            info[3] = flatReplace(info[3]);
            int t = (Integer.parseInt(info[1].substring(0, 2)) - Integer.parseInt(info[0].substring(0, 2))) * 60 + Integer.parseInt(info[1].substring(3, 5)) - Integer.parseInt(info[0].substring(3, 5));
            
            // 시작 멜로디가 같은 부분 찾기
            for (int i = 0; i < t - m.length() + 1; i++) {
                if (info[3].charAt(i % info[3].length()) == m.charAt(0)) {
                    // 다음 멜로디들이 같은지 확인
                    boolean flag = true;
                    for (int j = 1; j < m.length(); j++) {
                        if (info[3].charAt((i + j) % info[3].length()) != m.charAt(j)) {
                            flag = false;
                            break;
                        }
                    }
                    // 마지막 멜로디 다음 문자가 #이면 다른 음정
                    if (flag && maxT < t) {
                        answer = info[2];
                        maxT = t;
                        break;
                    }
                }
            }
        }
        
        return answer;
    }
}